File: Medium/200. Number of Islands/solution.cpp
class Solution {
public:
    
    void dfs(int i, int j, vector<vector<char>> &grid, vector<vector<bool>> &visited){
        if (i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size() || grid[i][j] == '0' || visited[i][j])  return;
        visited[i][j] = true;
        dfs(i - 1, j, grid, visited);
        dfs(i + 1, j, grid, visited);
        dfs(i, j - 1, grid, visited);
        dfs(i, j + 1, grid, visited);
    }
    
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<bool> > visited(m, vector<bool> (n, false));
        //vector<vector<int> > dir = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        
        int result = 0;
        
        for (int i = 0; i < m; i++){
            for (int j = 0; j < n; j++){
                if (grid[i][j] == '1' && !visited[i][j]){
                    dfs(i, j, grid, visited);
                    result++;
                }
            }
        }
        
        return result;
    }
};

File: Medium/200. Number of Islands/solution.py
class Solution:

    def fill(self, x, y, grid) -> None:
        grid[x][y] = '0'

        dx = [1, 0, -1, 0]
        dy = [0, 1, 0, -1]

        for i in range(4):
            nx = dx[i] + x
            ny = dy[i] + y

            if nx >= 0 and nx < len(grid) and ny >= 0 and ny < len(grid[0]) and grid[nx][ny] == '1':
                self.fill(nx, ny, grid)

    def numIslands(self, grid: List[List[str]]) -> int:
        cnt = 0

        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == '1':
                    cnt += 1
                    self.fill(i, j, grid)

        return cnt